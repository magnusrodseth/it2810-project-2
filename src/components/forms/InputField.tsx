import { InputHTMLAttributes } from 'react'
import { Input } from 'react-daisyui'

interface InputFieldProps extends Omit<InputHTMLAttributes<HTMLInputElement>, 'size' | 'color'> {
  label: string
  dataTestId?: string
}

const InputField: React.FC<InputFieldProps> = ({ label, dataTestId, ...props }) => {
  return (
    <div className="flex flex-col justify-center items-start mx-auto space-y-4">
      <label htmlFor={props.id} className="font-bold">{label}</label>
      <Input
        id={props.id}
        type="text"
        color="secondary"
        placeholder={props.placeholder}
        data-testid={dataTestId}
        required
        onChange={props.onChange}
        {...props}
      />
    </div>
  )
}

export default InputField
