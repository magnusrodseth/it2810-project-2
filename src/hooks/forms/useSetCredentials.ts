import { useEffect } from 'react'
import { SessionStorage } from '../../lib/storage'
import useRedirect from '../common/useRedirect'

/**
 * Sets credentials in session storage if the form input is valid and the user should log in.
 * @param isValid - is a boolean determining whether the form input is valid.
 * @param url - is the repository URL.
 * @param accessToken - is the access token.
 */
const useSetCredentials = (isValid: boolean, url: string, accessToken: string) => {
  useEffect(() => {
    if (isValid) {
      SessionStorage.set('repoUrl', url)
      SessionStorage.set('accessToken', accessToken)
      useRedirect('/dashboard')
    }
  }, [isValid])
}

export default useSetCredentials
