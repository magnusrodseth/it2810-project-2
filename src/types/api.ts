export type ApiCommitsOptions = {
  id: string
  withStats?: boolean
  perPage?: number
}

export type ApiIssuesOptions = {
  id: string
}

export type AuthenticationResponse = {
  ok: boolean
  message: string
}

export type CommitStats = {
  additions: number
  deletions: number
  total: number
}

export type Commit = {
  id: string
  short_id: string
  title: string
  author_name: string
  author_email: string
  authored_date: string
  committer_name: string
  committer_email: string
  committed_date: string
  created_at: string
  message: string
  parent_ids: string[]
  web_url: string
  stats?: CommitStats
}

export type Issue = {
  id: number
  issue_type: 'issue' | 'incident' | 'test_case'
  labels: string[]
  state: 'all' | 'opened' | 'closed'
}
