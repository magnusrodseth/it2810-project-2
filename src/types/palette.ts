import { Theme } from './theme'

export type Colors = {
  primary: string,
  secondary: string,
  accent: string,
  info: string,
  success: string,
  warning: string,
  error: string,
}

export type Palette = Record<Theme, Colors>
