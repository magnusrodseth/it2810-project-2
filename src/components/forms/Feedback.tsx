import { Severity } from '../../types/forms'
import classNames from '../../utils/common/classNames'

type FeedbackProps = {
  feedback: string
  severity: Severity
}

const Feedback: React.FC<FeedbackProps> = ({ feedback, severity }) => {
  return (
    <div
      data-testid="feedback"
      className={
        // Tailwind cannot interpolate the class, so it must be typed out manually
        classNames(
          'border-[1px] rounded-lg px-4 py-2 font-bold tracking-wide my-4',
          severity === 'error' ? 'border-error' : '',
          severity === 'warning' ? 'border-warning' : '',
          severity === 'info' ? 'border-info' : '',
          severity === 'success' ? 'border-success' : ''
        )
      }
    >
      <span>{`${feedback}.`}</span>
    </div>
  )
}

export default Feedback
