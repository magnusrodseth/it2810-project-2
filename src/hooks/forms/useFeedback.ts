import { useState } from 'react'
import { Severity } from '../../types/forms'

/**
 * Wrapper hook for providing feedback to the user with a given severity.
 *
 * Feedback is displayed as a result of submitting something, e.g. a form.
 * This feedback has a related severity, e.g. success, error, warning, etc.
 **/
const useFeedback = () => {
  const [feedback, _setFeedback] = useState<string | undefined>(undefined)
  const [severity, _setSeverity] = useState<Severity>('info')

  const setFeedback = (feedback: string, severity: Severity) => {
    _setFeedback(feedback)
    _setSeverity(severity)
  }

  return [feedback, severity, setFeedback] as const
}

export default useFeedback
