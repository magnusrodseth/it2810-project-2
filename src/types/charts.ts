import { CommitStats } from './api'

export type LineChartEntry = {
  timestamp: number
  commits: number
  author?: string
  stats?: CommitStats
}

export type BarChartEntry = {
  label: string
  count: number
}

export type LineChartData = LineChartEntry[]
export type BarChartData = BarChartEntry[]

export type YAxisLabel = keyof Pick<LineChartEntry, 'commits'>

export type Position = {
  left: string
  right: string
  top: string
  bottom: string
}

export type ChartType = 'commits' | 'issues'

export type Contributor = string

export type Option = {
  label: string
  value: string
}
