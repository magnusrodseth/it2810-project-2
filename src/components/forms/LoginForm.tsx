import Button from '../../components/common/Button'
import { useState } from 'react'
import { isAuthenticated } from '../../api/auth'
import useFeedback from '../../hooks/forms/useFeedback'
import { mapStatusToMessage } from '../../utils/api/handlers'
import { AxiosError } from 'axios'
import { AuthenticationResponse } from '../../types/api'
import Feedback from './Feedback'
import classNames from '../../utils/common/classNames'
import InputField from './InputField'
import { useRedirectIfAuthenticated } from '../../hooks/auth/useRedirectIfAuthenticated'
import isValidProjectUrl from '../../utils/validators/isValidProjectUrl'
import useSetCredentials from '../../hooks/forms/useSetCredentials'

const LoginForm = () => {
  const [feedback, severity, setFeedback] = useFeedback()
  const [url, setUrl] = useState('')
  const [accessToken, setAccessToken] = useState('')

  useRedirectIfAuthenticated()

  const handleLogin = async () => {
    if (!isValidProjectUrl(url)) {
      setFeedback('Invalid project URL', 'error')
      return
    }

    await isAuthenticated(url, accessToken)
      .then((response) => {
        setFeedback(response.message, 'success')
      })
      .catch((error: AxiosError<AuthenticationResponse>) => {
        const feedback =
          error.response?.data.message || mapStatusToMessage(error.response?.status || 500)

        setFeedback(feedback, 'error')
      })
  }

  const isValid = severity === 'success'
  useSetCredentials(isValid, url, accessToken)

  return (
    <div>
      {feedback && (
        <div className="my-4 text-center">
          <Feedback feedback={feedback} severity={severity} />
        </div>
      )}

      <form
        className={classNames(
          'flex flex-col justify-center',
          'rounded-lg rise-on-hover bg-base-300 shadow-md hover:shadow-lg',
          'space-y-8 p-6 md:p-10'
        )}
        onSubmit={(event) => {
          event.preventDefault()
          handleLogin()
        }}>
        <InputField
          label="GitLab Repository URL"
          id="input-url"
          placeholder="Type here"
          dataTestId="input-url"
          onChange={(event) => setUrl(event.target.value)}
          required
        />

        <InputField
          label="Personal access token"
          id="input-access-token"
          placeholder="Type here"
          dataTestId="input-access-token"
          onChange={(event) => setAccessToken(event.target.value)}
          required
        />

        <Button color="secondary" type="submit" className="mt-6">
          Sign In
        </Button>
      </form>
    </div>
  )
}

export default LoginForm
