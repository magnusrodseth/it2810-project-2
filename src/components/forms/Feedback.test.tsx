import { render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { Severity } from '../../types/forms'
import Feedback from './Feedback'

const assertCorrectStyle = (severity: Severity, className: string) => {
  render(<Feedback feedback={'feedback'} severity={severity} />)
  const component = screen.getByTestId('feedback')
  expect(component).toHaveClass(className)
}

describe('Feedback', () => {
  it('renders correctly', () => {
    const component = renderer.create(<Feedback feedback={'feedback'} severity={'success'} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders error style with "error" severity', () => {
    assertCorrectStyle('error', 'border-error')
  })

  it('renders warning style with "warning" severity', () => {
    assertCorrectStyle('warning', 'border-warning')
  })

  it('renders info style with "info" severity', () => {
    assertCorrectStyle('info', 'border-info')
  })

  it('renders success style with "success" severity', () => {
    assertCorrectStyle('success', 'border-success')
  })
})

export {}
