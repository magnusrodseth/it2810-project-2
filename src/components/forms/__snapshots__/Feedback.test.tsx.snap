// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback renders correctly 1`] = `
<div
  className="border-[1px] rounded-lg px-4 py-2 font-bold tracking-wide my-4 border-success"
  data-testid="feedback"
>
  <span>
    feedback.
  </span>
</div>
`;
